design/one-pager-package-image-config.md

```diff
@@ -0,0 +1,377 @@
+TLS settings. It also provides a way to define policies for verifying image
+signatures.
+
+Considering the API is solving some existing problems, e.g. configuring pull
+secrets for dependencies, and not introducing a behavioral change unless the
+users opt in by creating an `ImageConfig` object, we believe this API should be
+introduced as a beta API without going through the feature lifecycle stages.
+
+### API
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: ImageConfig
+metadata:
+  name: acme-packages
+spec:
+  matchImages:
+    - prefix: registry1.com/acme-co/configuration-foo
+    - prefix: registry1.com/acme-co/configuration-bar
+    - prefix: registry1.com/acme-co/function-baz
+  registry:
+    authentication:
+      pullSecretRef:
+        name: acme-registry-credentials
+    tls:
+      enabled: true
+      insecure: false
+      caBundleConfigMapRef:
+        name: registry1-ca
+        key: ca.crt
+  verification:
+    provider: cosign
+    cosign:
+      authorities:
+        - name: verify acme builds
+          keyless:
+            url: https://fulcio.sigstore.dev
+            identities:
+              - issuer: https://token.actions.githubusercontent.com
+                subjectRegExp: https://github.com/acme-co/crossplane-packages/*
+          attestations:
+            - name: verify attestations
+              predicateType: spdxjson
+```
+
+The `ImageConfig` API `spec` has the following fields:
+
+- `matchImages`: A list of image prefixes to match. The package manager will
+  apply the settings defined in the `ImageConfig` object to the images that
+  match the prefixes. The longest prefix match will be selected in case of
+  multiple matches.
+- `registry`: Configuration for interacting with the registry hosting the images.
+  - `authentication`: Authentication settings for the registry.
+    - `pullSecretRef`: Reference to the Kubernetes Secret containing the
+      credentials to pull images from the registry. This secret must be of type
+      [`kubernetes.io/dockerconfigjson`].
+  - `tls`: TLS settings for the registry.
+    - `enabled`: Whether to enable TLS for the registry. Default is `true`.
+    - `insecure`: Whether to skip verifying the server's certificate. Default is
+      `false`.
+    - `caBundleConfigMapRef`: Reference to the ConfigMap containing the CA
+      certificate bundle to use for verifying the server's certificate.
+- `verification`: Configuration for verifying image signatures.
+  - `provider`: The provider to use for verifying image signatures. In the
+    beginning, only `cosign` will be supported.
+  - `cosign`: Configuration for verifying images using cosign.
+    - `authorities`: List of authorities to use for verifying images.
+      - `name`: The name of the authority.
+      - `keyless`: Keyless verification settings.
+        - `url`: The URL of the keyless authority.
+        - `identities`: List of identities to use for verification.
```

Comment:
I guess these identities will be AND-ed? So all identities need to have signed the image for the image to be accepted?

Reply:
It looks like it can: [Multiple signatures can be embedded in one image manifest.](https://github.com/sigstore/cosign/blob/a0752eb40b500316ac417baf4926a2c2d99b39b8/specs/SIGNATURE_SPEC.md?plain=1#L184C1-L184C59)
And it used to be common practice with Notary/TUF to do so during pipelines, multiple stages could have different identities and only if an image is signed by all of them it's ok to use in production. I'm not up to speed with cosign practices though.

Reply:
also kyverno [docs](https://kyverno.io/docs/writing-policies/verify-images/sigstore/#verifying-image-signatures) say so: `An image can be signed with multiple signatures, for example at the organization level and at the project level.`

Reply:
I would rather stick to policy controller / cosign practices, and I am planning to leverage libraries/code there.